File: bulk.js
import 'dotenv/config';
import { createClient } from '@supabase/supabase-js';
import XLSX from 'xlsx';

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_SERVICE_ROLE_KEY
);

// -------------------- CONFIG --------------------
const FILE = './college.xlsx';
const STAFF_SHEET = 'Sheet1';     // CHANGE if needed
const TIMETABLE_SHEET = 'Sheet1'; // same sheet in your case

const DAY_MAP = {
  M: 'monday',
  T: 'tuesday',
  W: 'wednesday',
  Th: 'thursday',
  F: 'friday',
  Sa: 'saturday',
};

function emptyWeek() {
  return {
    monday: Array(7).fill(null),
    tuesday: Array(7).fill(null),
    wednesday: Array(7).fill(null),
    thursday: Array(7).fill(null),
    friday: Array(7).fill(null),
    saturday: Array(7).fill(null),
  };
}

// -------------------- LOAD EXCEL --------------------
const workbook = XLSX.readFile(FILE);
const sheet = XLSX.utils.sheet_to_json(workbook.Sheets[STAFF_SHEET]);

if (!sheet.length) {
  throw new Error('Excel sheet empty or wrong sheet name');
}

// -------------------- MAIN --------------------
async function run() {
  for (const row of sheet) {
    const email = row['Mail id']?.trim();
    const name = row['Staff Name']?.trim();

    if (!email || !name) {
      console.warn('⏭ Skipping row (missing email/name)');
      continue;
    }

    console.log(`\n▶ Processing ${email}`);

    // -----------------------------------
    // 1. CREATE OR FETCH AUTH USER
    // -----------------------------------
    let userId;

    const { data: existing } = await supabase.auth.admin.listUsers();
    const found = existing.users.find(u => u.email === email);

    if (found) {
      userId = found.id;
      console.log('ℹ Auth user exists');
    } else {
      const { data, error } = await supabase.auth.admin.createUser({
        email,
        email_confirm: true,
        user_metadata: { full_name: name },
      });

      if (error) {
        console.error('❌ Auth error:', error.message);
        continue;
      }

      userId = data.user.id;
      console.log('✔ Auth user created');
    }

    // -----------------------------------
    // 2. WAIT FOR STAFF TRIGGER
    // -----------------------------------
    let staffId = null;
    for (let i = 0; i < 10; i++) {
      const { data } = await supabase
        .from('staff')
        .select('id')
        .eq('profile_id', userId)
        .maybeSingle();

      if (data?.id) {
        staffId = data.id;
        break;
      }
      await new Promise(r => setTimeout(r, 500));
    }

    if (!staffId) {
      console.error('❌ Staff row not found for', email);
      continue;
    }

    // -----------------------------------
    // 3. BUILD TIMETABLE
    // -----------------------------------
    const week = emptyWeek();

    for (const key of Object.keys(row)) {
      const m = key.match(/^(M|T|W|Th|F|Sa)(\d)$/);
      if (!m) continue;

      const day = DAY_MAP[m[1]];
      const idx = Number(m[2]) - 1;

      if (row[key]) {
        week[day][idx] = row[key].toString().trim();
      }
    }

    // -----------------------------------
    // 4. UPSERT TIMETABLE
    // -----------------------------------
    const { error: ttErr } = await supabase
      .from('timetable')
      .upsert({
        staff_id: staffId,
        ...week,
      }, { onConflict: 'staff_id' });

    if (ttErr) {
      console.error('❌ Timetable error:', ttErr.message);
      continue;
    }

    console.log('✔ Timetable saved');
  }

  console.log('\n🎉 IMPORT COMPLETED SUCCESSFULLY');
}

run().catch(err => {
  console.error('🔥 FATAL:', err);
});

File: bulk.py
import os
import re
import pandas as pd
from supabase import create_client
from dotenv import load_dotenv
from openpyxl import load_workbook

# ================= LOAD ENV =================
load_dotenv()

SUPABASE_URL = os.getenv("SUPABASE_URL")
SUPABASE_SERVICE_ROLE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY")

supabase = create_client(SUPABASE_URL, SUPABASE_SERVICE_ROLE_KEY)

# ================= CONFIG =================
FILE = "cse.xlsx"
BUCKET = "avatars"
DEPT = "CSE"

DAY_MAP = {
    "M": "monday",
    "T": "tuesday",
    "W": "wednesday",
    "Th": "thursday",
    "F": "friday",
    "Sa": "saturday",
}

def empty_week():
    return {day: [None] * 7 for day in DAY_MAP.values()}

# ================= LOAD AUTH USERS =================
print("📦 Loading auth users...")
auth_res = supabase.auth.admin.list_users(per_page=1000)
auth_users = {u.email: u for u in auth_res}

# ================= EXTRACT IMAGES =================
print("🖼 Extracting images from Excel...")
wb = load_workbook(FILE)
ws = wb.active

row_images = {}
for img in ws._images:
    excel_row = img.anchor._from.row + 1
    row_images[excel_row] = img._data()

# ================= LOAD EXCEL DATA =================
df = pd.read_excel(FILE)

# ================= MAIN LOOP =================
for index, row in df.iterrows():
    excel_row = index + 2  # header offset
    email = str(row.get("Mail id", "")).strip()
    name = str(row.get("Staff Name", "")).strip()

    if not email or email.lower() == "nan":
        continue

    print(f"\n▶ Processing {email}")

    # ---------- AUTH (CREATE OR UPDATE) ----------
    user = auth_users.get(email)

    if not user:
        res = supabase.auth.admin.create_user({
            "email": email,
            "email_confirm": True,
            "user_metadata": {"full_name": name},
        })
        user = res.user
        auth_users[email] = user
        print("👤 Auth user created")
    else:
        supabase.auth.admin.update_user_by_id(
            user.id,
            {"user_metadata": {"full_name": name}},
        )
        print("🔄 Auth user exists → metadata updated")

    user_id = user.id

    # ---------- STAFF UPSERT ----------
    staff_payload = {
        "profile_id": user_id,
        "dept": DEPT,
    }

    # ---------- IMAGE: DELETE → UPLOAD ----------
    if excel_row in row_images:
        file_path = f"{user_id}/profile.jpg"

        try:
            supabase.storage.from_(BUCKET).remove([file_path])
            print("🗑 Old photo removed (if existed)")
        except Exception:
            pass

        supabase.storage.from_(BUCKET).upload(
            file_path,
            row_images[excel_row],
            file_options={
                "content-type": "image/jpeg",
                "upsert": "false",
            },
        )

        staff_payload["photo_url"] = (
            f"{SUPABASE_URL}/storage/v1/object/public/{BUCKET}/{file_path}"
        )

        print("✔ Photo uploaded")

    supabase.table("staff").upsert(
        staff_payload,
        on_conflict="profile_id",
    ).execute()

    # Fetch staff_id safely
    staff_row = (
        supabase.table("staff")
        .select("id")
        .eq("profile_id", user_id)
        .single()
        .execute()
    )

    staff_id = staff_row.data["id"]
    print("✔ Staff upserted")

    # ---------- TIMETABLE UPSERT ----------
    week = empty_week()

    for col in df.columns:
        match = re.match(r"^(M|T|W|Th|F|Sa)(\d)$", str(col))
        if match:
            day = DAY_MAP[match.group(1)]
            idx = int(match.group(2)) - 1
            value = row.get(col)
            if pd.notna(value):
                week[day][idx] = str(value).strip()

    supabase.table("timetable").upsert(
        {"staff_id": staff_id, **week},
        on_conflict="staff_id",
    ).execute()

    print("✔ Timetable upserted")

print("\n🎉 IMPORT COMPLETED — SAFE TO RE-RUN")
